##### This notebook will be used for the Coursera Capstone project.

In [1]:
#import libraries
import pandas as pd
import numpy as np
print ("Hello Capstone Project Course!")

Hello Capstone Project Course!


## 1. Introduction/Business Problem

In today's world many people can't imagine their lives without a car. Cars have become an integral part of modern humans, making it easier not only to perform everyday activities but also to travel long distance. Nonetheless, road vehicles are also the most unsafe of all available to humans. Traffic collisions are in first place in terms of the number of deaths and injuries. According to these parameters, cars significantly overtake railway, aviation and water transport.

Unfortunately, car accidents occur for many reasons, including both technological and human factors. An accident can happen due to the fault of a tired driver, due to icing of the road surface or a malfunction of the brake system. However, the risk of getting into an accident is often influenced by external factors, such as the day of the week, weather conditions and the quality of the road itself.

Many car companies are working on fully autonomous cars, which could potentially reduce the amount of car accidents. Until then regular cars will remain popular. Therefore, it would be great to have a mechanism that could warn you, given the weather and the road conditions, about the possibility of you getting into a car accident and how severe it would be, so that you would drive more carefully or even change your travel if you are able to. 



Thus, the goal of this project is:
#### 1.	to identify the factors that significantly affect the risk of traffic collisions, and
#### 2.	to create a model that will predict the severity of car accidents.

The findings may be useful for improving road safety or for insurance companies planning to introduce life and health insurance programs for drivers and passengers. 



## 2. Data

For this project I will be using a shared dataframe called "Collisions" that is provided by the IBM. 

The data in the dataset is labeled, which will allow me to use a supervised machine-learning model. 

In [2]:
#download data
!wget -O DataCollisions.csv https://s3.us.cloud-object-storage.appdomain.cloud/cf-courses-data/CognitiveClass/DP0701EN/version-2/Data-Collisions.csv

--2020-10-03 21:22:08--  https://s3.us.cloud-object-storage.appdomain.cloud/cf-courses-data/CognitiveClass/DP0701EN/version-2/Data-Collisions.csv
Resolving s3.us.cloud-object-storage.appdomain.cloud (s3.us.cloud-object-storage.appdomain.cloud)... 67.228.254.196
Connecting to s3.us.cloud-object-storage.appdomain.cloud (s3.us.cloud-object-storage.appdomain.cloud)|67.228.254.196|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 73917638 (70M) [text/csv]
Saving to: ‘DataCollisions.csv’

DataCollisions.csv  100%[===================>]  70.49M  19.6MB/s    in 3.7s    

2020-10-03 21:22:12 (18.9 MB/s) - ‘DataCollisions.csv’ saved [73917638/73917638]



In [3]:
#read the online file by the URL provided above, and assign it to variable "df"
df = pd.read_csv('DataCollisions.csv') 

#check the top n rows of the dataframe
df.head() 

/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/IPython/core/interactiveshell.py:3072: DtypeWarning: Columns (33) have mixed types.Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


SEVERITYCODE           X          Y  OBJECTID  INCKEY  COLDETKEY REPORTNO  \
0             2 -122.323148  47.703140         1    1307       1307  3502005   
1             1 -122.347294  47.647172         2   52200      52200  2607959   
2             1 -122.334540  47.607871         3   26700      26700  1482393   
3             1 -122.334803  47.604803         4    1144       1144  3503937   
4             2 -122.306426  47.545739         5   17700      17700  1807429   

    STATUS      ADDRTYPE   INTKEY  ... ROADCOND                LIGHTCOND  \
0  Matched  Intersection  37475.0  ...      Wet                 Daylight   
1  Matched         Block      NaN  ...      Wet  Dark - Street Lights On   
2  Matched         Block      NaN  ...      Dry                 Daylight   
3  Matched         Block      NaN  ...      Dry                 Daylight   
4  Matched  Intersection  34387.0  ...      Wet                 Daylight   

  PEDROWNOTGRNT  SDOTCOLNUM SPEEDING ST_COLCODE  \
0           NaN         NaN      NaN         10   
1           NaN   6354039.0      NaN         11   
2           NaN   4323031.0      NaN         32   
3           NaN         NaN      NaN         23   
4           NaN   4028032.0      NaN         10   

                                          ST_COLDESC  SEGLANEKEY  \
0                                  Entering at angle           0   
1  From same direction - both going straight - bo...           0   
2                             One parked--one moving           0   
3                   From same direction - all others           0   
4                                  Entering at angle           0   

   CROSSWALKKEY  HITPARKEDCAR  
0             0             N  
1             0             N  
2             0             N  
3             0             N  
4             0             N  

[5 rows x 38 columns]

The data is about “accident severity”. The dataset includes all types of collisions for the timeframe from 2004 to 2018. 

The dataset has an extensive amount of observations - more than 190,000. 

The total amount of attributes in the dataframe is 37. 

In [4]:
#check size of dataframe (rows, columns)
df.shape

(194673, 38)

However, not all attributes will be useful for this project. 

In [5]:
#check the name of columns(headers) 
df.columns 

Index(['SEVERITYCODE', 'X', 'Y', 'OBJECTID', 'INCKEY', 'COLDETKEY', 'REPORTNO',
       'STATUS', 'ADDRTYPE', 'INTKEY', 'LOCATION', 'EXCEPTRSNCODE',
       'EXCEPTRSNDESC', 'SEVERITYCODE.1', 'SEVERITYDESC', 'COLLISIONTYPE',
       'PERSONCOUNT', 'PEDCOUNT', 'PEDCYLCOUNT', 'VEHCOUNT', 'INCDATE',
       'INCDTTM', 'JUNCTIONTYPE', 'SDOT_COLCODE', 'SDOT_COLDESC',
       'INATTENTIONIND', 'UNDERINFL', 'WEATHER', 'ROADCOND', 'LIGHTCOND',
       'PEDROWNOTGRNT', 'SDOTCOLNUM', 'SPEEDING', 'ST_COLCODE', 'ST_COLDESC',
       'SEGLANEKEY', 'CROSSWALKKEY', 'HITPARKEDCAR'],
      dtype='object')

Also, some attributes include missing data.

In [6]:
#check the number of missing values in each column
df.isnull().sum()

SEVERITYCODE           0
X                   5334
Y                   5334
OBJECTID               0
INCKEY                 0
COLDETKEY              0
REPORTNO               0
STATUS                 0
ADDRTYPE            1926
INTKEY            129603
LOCATION            2677
EXCEPTRSNCODE     109862
EXCEPTRSNDESC     189035
SEVERITYCODE.1         0
SEVERITYDESC           0
COLLISIONTYPE       4904
PERSONCOUNT            0
PEDCOUNT               0
PEDCYLCOUNT            0
VEHCOUNT               0
INCDATE                0
INCDTTM                0
JUNCTIONTYPE        6329
SDOT_COLCODE           0
SDOT_COLDESC           0
INATTENTIONIND    164868
UNDERINFL           4884
WEATHER             5081
ROADCOND            5012
LIGHTCOND           5170
PEDROWNOTGRNT     190006
SDOTCOLNUM         79737
SPEEDING          185340
ST_COLCODE            18
ST_COLDESC          4904
SEGLANEKEY             0
CROSSWALKKEY           0
HITPARKEDCAR           0
dtype: int64

In [7]:
#drop the unnecessary columns
#use pandas .drop() method 
#in pandas axis=0 represents rows (default) and axis=1 represents columns

df.drop(['X', 'Y', 'OBJECTID', 'INCKEY', 'COLDETKEY', 'REPORTNO',
       'STATUS', 'INTKEY', 'LOCATION', 'EXCEPTRSNCODE',
       'EXCEPTRSNDESC', 'SEVERITYCODE.1', 'SEVERITYDESC', 'COLLISIONTYPE',
       'PEDCOUNT', 'PEDCYLCOUNT',
       'INCDTTM', 'JUNCTIONTYPE', 'SDOT_COLCODE', 'SDOT_COLDESC',
       'INATTENTIONIND', 'UNDERINFL',
       'PEDROWNOTGRNT', 'SDOTCOLNUM', 'ST_COLCODE', 'ST_COLDESC',
       'SEGLANEKEY', 'CROSSWALKKEY', 'HITPARKEDCAR'], axis=1, inplace=True)
df.head()

SEVERITYCODE      ADDRTYPE  PERSONCOUNT  VEHCOUNT                 INCDATE  \
0             2  Intersection            2         2  2013/03/27 00:00:00+00   
1             1         Block            2         2  2006/12/20 00:00:00+00   
2             1         Block            4         3  2004/11/18 00:00:00+00   
3             1         Block            3         3  2013/03/29 00:00:00+00   
4             2  Intersection            2         2  2004/01/28 00:00:00+00   

    WEATHER ROADCOND                LIGHTCOND SPEEDING  
0  Overcast      Wet                 Daylight      NaN  
1   Raining      Wet  Dark - Street Lights On      NaN  
2  Overcast      Dry                 Daylight      NaN  
3     Clear      Dry                 Daylight      NaN  
4   Raining      Wet                 Daylight      NaN

For this project the following attributes were chosen:
* 'SEVERITYCODE' - Target variable, which is used to measure the severity of an accident from 0 to 4 within the dataset;
* 'ADDRTYPE' - Collision address type (Alley, Block, Intersection);
* 'PERSONCOUNT' - The number of pedestrians involved in the collision;
* 'VEHCOUNT' - The number of vehicles involved in the collision;
* 'INCDATE' - The date of the incident;
* 'WEATHER' - A description of the weather conditions during the time of the collision;
* 'ROADCOND' - The condition of the road during the collision;
* 'LIGHTCOND' - The light conditions during the collision;
* 'SPEEDING' - Whether or not speeding was a factor in the collision.


In [8]:
#check the number of missing values in each column
df.isnull().sum()

SEVERITYCODE         0
ADDRTYPE          1926
PERSONCOUNT          0
VEHCOUNT             0
INCDATE              0
WEATHER           5081
ROADCOND          5012
LIGHTCOND         5170
SPEEDING        185340
dtype: int64

**Note**: the attribute “SPEEDING” has more than 185,000 missing values, which can create a biased ML model. Therefore, this attribute wwill be dropped.

Rows with missing values from the attributes "ADDRTYPE","WEATHER","ROADCOND","LIGHTCOND" were also dropped.

In [9]:
#drop the attribute “SPEEDING”
df.drop(['SPEEDING'], axis=1, inplace=True)

In [10]:
#drop all rows with missing values
df.dropna(subset=["ADDRTYPE","WEATHER","ROADCOND","LIGHTCOND"], axis=0, inplace=True)

#reset index, because we droped the rows
df.reset_index(drop=True, inplace=True)

In [11]:
#summary of a dataframe
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 187525 entries, 0 to 187524
Data columns (total 8 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   SEVERITYCODE  187525 non-null  int64 
 1   ADDRTYPE      187525 non-null  object
 2   PERSONCOUNT   187525 non-null  int64 
 3   VEHCOUNT      187525 non-null  int64 
 4   INCDATE       187525 non-null  object
 5   WEATHER       187525 non-null  object
 6   ROADCOND      187525 non-null  object
 7   LIGHTCOND     187525 non-null  object
dtypes: int64(3), object(5)
memory usage: 11.4+ MB


Now, we have finally obtained the cleaned dataset with no missing values and all data in its proper format. 

The final size of the dataframe includes 187524 entries and 8 attributes.

In [12]:
df

SEVERITYCODE      ADDRTYPE  PERSONCOUNT  VEHCOUNT  \
0                  2  Intersection            2         2   
1                  1         Block            2         2   
2                  1         Block            4         3   
3                  1         Block            3         3   
4                  2  Intersection            2         2   
...              ...           ...          ...       ...   
187520             2         Block            3         2   
187521             1         Block            2         2   
187522             2  Intersection            3         2   
187523             2  Intersection            2         1   
187524             1         Block            2         2   

                       INCDATE   WEATHER ROADCOND                LIGHTCOND  
0       2013/03/27 00:00:00+00  Overcast      Wet                 Daylight  
1       2006/12/20 00:00:00+00   Raining      Wet  Dark - Street Lights On  
2       2004/11/18 00:00:00+00  Overcast      Dry                 Daylight  
3       2013/03/29 00:00:00+00     Clear      Dry                 Daylight  
4       2004/01/28 00:00:00+00   Raining      Wet                 Daylight  
...                        ...       ...      ...                      ...  
187520  2018/11/12 00:00:00+00     Clear      Dry                 Daylight  
187521  2018/12/18 00:00:00+00   Raining      Wet                 Daylight  
187522  2019/01/19 00:00:00+00     Clear      Dry                 Daylight  
187523  2019/01/15 00:00:00+00     Clear      Dry                     Dusk  
187524  2018/11/30 00:00:00+00     Clear      Wet                 Daylight  

[187525 rows x 8 columns]